Adds framework for status effects

Xhuis opened the pull request:
Added framework for status effects with lots of customization, including duration, effects, and pretty much everything else. Files are located in `__DEFINES/status_effects.dm`, and `datum/status_effects`. There are two procs for adding and removing them, both at the mob level - `apply_status_effect` and `remove_status_effect`, both of which take each effect's path or define.

This allows for making timed or constant effects without having random loops all over the place. For instance, if you want to have a creature mark its target to die after one minute, you can make a status effect that lasts one minute and calls `death()` on the owner once it expires.

Code suggestions and criticisms are appreciated and encouraged.

Commit:
Adds framework for status effects

Commit:
Completes the Renew define

Xhuis commented on code/datums/status_effects/status_effect.dm:
```diff
@@ -0,0 +1,52 @@
+
+/datum/status_effect/proc/start_ticking()
+	on_apply()
+	while(TRUE) //please don't kill me :c
```
I found that this increased the flexibility I have, and it shouldn't have any recursive effects due to it being an individual datum.

RemieRichards replied:
it should just be something that processes, not custom snowflake.

Xhuis replied:
I considered that. Would a new subsystem be worth making? I used SSobj but its delay for processing really hindered me.

RemieRichards replied:
just put it on the fast processing SS?

Xhuis replied:
Could be doable. Fastprocess ticks five times per second, right?

optimumtact replied:
We're very unlikely to merge this when it reimplements the process scheduling and subsystem ticker we already have for it's own benefit.

AnturK replied:
If this is meant to be standard effect system separate controller is reasonable.

RemieRichards commented on code/datums/status_effects/status_effect.dm:
```diff
@@ -0,0 +1,52 @@
+	S.owner = src
+
+/mob/living/proc/remove_status_effect(effect)
+	for(var/datum/status_effect/S in status_effects)
```
so you can have multiples of the same buff? that's... weird.
Like sure, buff-stacking to get a better version of the buff, but just plain having two copies? weird.

Xhuis replied:
That's an oversight! I'll make a variable for it.

optimumtact commented:
Are you planning on moving the fire and atmos status effects to this system?

i.e being set on fire applies the burning effect (alert and damage) and losing air or breathing poison applies the atmos alerts and associated damage?

It's good to have a system as in the past people have often used the chem or disease systems to apply these effects instead.

If someone was enterprising they could also overhaul disease symptoms to be effects, so diseases would be a layer above this system dealing with the virology aspects.

Xhuis commented:
Disease already has its own datum thingy, so I probably won't change that. On fire effects would be easy enough to do, but I think this will be a more niche system.

optimumtact commented:
Why would it be niche? There are many items in the game that need to apply a consistent/timed status effect on the bearer or on some other condition.

Xhuis commented:
I'm sure there's plenty of ways this can be applied. I just don't know if things like atmos and viruses would be suitable candidates as they already have lots of code written around them. I don't want _really_ complex stuff - only small stuff like "take this amount of damage over this amount of time" and stuff like that. If you can list off stuff that I can refactor to status effects, I'd definitely refactor them... tomorrow. Sleep now.

WJohn commented:
Weaken and fire and weightlessness and sleep or whatever should totally be moved to status effect with alerts. More obvious feedback for what's going on.

AnturK commented:
Good idea but needs to actually standarize current statuses. Also ideally link it with the popup alerts.

Commit:
Lots of work, refactoring

Xhuis commented:
* Moved Invoke Inath-neq to a status effect, Inath-neq's Endowment. More refactors to come.
* Made status effects throw alerts to the afflicted' screen, based on icons found in `status_effects.dmi`.
* Refactored fastprocess into a processing subsystem that fires once per second. Fastprocess is a child of this that fires at its previous rate.
* Added support for non-stacking status effects, which all status effects will utilize by default.
* Moved all processing subsystems into `/subsystems/processing/`.

optimumtact commented:
Ready for review?

Xhuis commented:
Yeah.

RemieRichards commented on code/controllers/subsystem/processing/processing.dm:
```diff
@@ -0,0 +1,34 @@
+		if(thing)
+			thing.process(wait)
+		else
+			SSfastprocess.processing -= thing
```
error?

Xhuis replied:
Yes, thank you

RemieRichards commented on code/datums/status_effects/status_effect.dm:
```diff
@@ -0,0 +1,68 @@
+
+/mob/living/proc/apply_status_effect(effect)
+	var/datum/status_effect/S = new effect
+	for(var/datum/status_effect/S2 in status_effects)
```
you could just make ```status_effects``` an assoc list of ```type = status_effect instance```, that prevents dupes, is faster to access, and if you choose allows setting an effect again to essentially just "refresh" it.

Xhuis replied:
So each status effect would exist at roundstart and be "assigned" to mobs who lose it after the duration, but the effect still exists? I suppose that could work.

RemieRichards replied:
the list wouldn't have to start with anything. if something tried to remove a non-existent effect, there'd be no issues, no boom, no runtime explosion. the list would start empty, but would slowly be filled like an assoc list.

it's all just better code-bookkeeping.

Xhuis replied:
Ah, I see what you mean! Good point.

Xhuis replied:
I've been working on this but I'm pretty atrocious with associative list syntax and I wanted to know what the problem with this would be.

Here's what I have for the `New()` for status effects:
```
/datum/status_effect/New()
	..()
	spawn(1) //Give us time to set any variables
		start_ticking()
		status_effects += src["[owner.real_name][id]"]
```
But I can't help but feel that this is wrong. Am I missing something?

RemieRichards replied:
yes that is indeed wrong.
```status_effects``` is the assoc list, so you do the list operation on IT.
```status_effects["[owner.real_name][id]"] = src```, and it's ```=``` because you're setting the value of ```status_effects["[owner.real_name][id]"]```, as opposed to adding to that value (if that value was itself a list, for example.

tl;dr
```
list[1] = "A" //First element is now "A"
assoc_list["some_key"] = "A" //the key "some_key" now points to "A"
```

Xhuis replied:
Thanks. I'll post again if I need further help.

Xhuis replied:
Hey, reporting back in. I've refactored it for the most part, but there's a small obstacle I can't surmount. It's with this code:
```
/mob/living/proc/apply_status_effect(datum/status_effect/effect)
	if(status_effects["[real_name][initial(effect.id)]"])
		return 0
	var/datum/status_effect/S = new effect
	S.owner = src
	return 1
```
It works fine, but removing won't. Is typecasting `effect` as an actual status effect correct, and will `initial(effect.id)` work in this context?

RemieRichards replied:
yes, typecasting like that works fine.
it's a little dangerous because you're not checking if effect is a path or not, but it works.

Xhuis replied:
All right. My issue here is that `apply_status_effect` always returns 1, meaning that it never picks up the status effect in question. This is included in the `New()` for status effects now.
`status_effects["[owner.real_name][id]"] = src`
So I'm not entirely sure what's going wrong. I think it's because `apply_status_effect` takes a path, not an actual object. Would it be better to just make a new status effect for reference and delete it once the code is done with it?

RemieRichards replied:
@Xhuis you're accessing the id from the path using ```initial()``` so it doesn't matter that it's not an object.

does your ```apply_status_effect()``` look exactly as that comment? because you're setting S's owner AFTER ```New()```, when you just said you put the assoc list addition in ```New()```

Xhuis replied:
Still having issues after some tweaks, so I'll post the code to hear your own thoughts. Here's everything relevant.

```
/datum/status_effect/New(loc, mob/living/initial_owner)
	..()
	if(initial_owner)
		owner = initial_owner
	status_effects["[owner.real_name][id]"] = src
	start_ticking()

/datum/status_effect/Destroy()
	status_effects["[owner.real_name][id]"] = null
	..()

/mob/living/proc/apply_status_effect(datum/status_effect/effect)
    if(status_effects["[real_name][initial(effect.id)]"])
        return 0
    var/datum/status_effect/S = new effect
    S.owner = src
    return 1

/mob/living/proc/remove_status_effect(datum/status_effect/effect)
	var/datum/status_effect/S = status_effects["[real_name][initial(effect.id)]"]
	if(!S)
		return 0
	world << S.name
	qdel(S)
	return 1
```

RemieRichards replied:
Here's a better ```New()```:
```
/datum/status_effect/New(new_owner)
    ..()
    if(new_owner)
        owner = new_owner
        status_effects["[owner.real_name][id]"] = src
        start_ticking()
    else
        qdel(src)
```
Better ```Destroy()```:
```
/datum/status_effect/Destroy()
    status_effects -= "[owner.real_name][id]" 
    return ..() //Destroy()s MUST return their hint.
```

Xhuis replied:
Thanks again for the help. I'll stay in touch.

Xhuis replied:
Afraid there's no change. It still adds duplicate effects and fails to remove existing ones.

RemieRichards replied:
use the person's ```\ref[]``` (```\ref[src]```) instead of their ```real_name```

Xhuis replied:
That just throws a bunch of undefined variable errors.

Xhuis replied:
Losing motivation because this _still_ won't work. Reverting back to the old system.

RemieRichards replied:
are you sure you used ```\ref[src]``` right? it's ```"some text blah blah \ref[a_var_such_as_src]"```, where as I'm assuming you thought to add ```[]``` around the whole statement, which is wrong.

Xhuis replied:
So the syntax would be:
`status_effects[\ref[owner][id]]`?

RemieRichards replied:
```status_effects["\ref[owner][id]"]```

Xhuis replied:
Through debugging, I discovered that `initial(effect.id)` when referring to the path comes up null. I think it _does_ have to be called on an object.

RemieRichards replied:
that makes no sense, as it's never null, at worst it's "effect"

Xhuis replied:
Yeah, I don't know either. I think the answer boils down to "Byond".

optimumtact replied:
why are you doing intial(id)?

optimumtact commented:
Any chance of working on those suggestions this week @Xhuis ?

Xhuis commented:
Sure. I'll do some work when I get the opportunity.

Commit:
Further work on status effects

optimumtact commented:
Is this ready to go?

Xhuis commented:
No. It still doesn't work and I'm trying to figure out the best solution for the problem it's having.

Razharas commented:
Make crit into a status effect

Xhuis commented:
@Razharas That'd be for another PR since it would involve changing the fundamentals of human `Life` code.

Anyway, we have two options with this PR;
1. Use the inferior, slower, and clunkier method and merge it since the other method doesn't work
2. Close it or put it on hiatus until I can figure out _why_ it doesn't work

I've tried getting the ideal method to work for almost a week now, to no avail. What do you think, maintainers?

optimumtact commented:
I wouldn't want to hold up what is a good overhaul for just that, so go back to the slower method for now and we can look at fixing it later.

Xhuis commented:
All right, once I finish that highlander PR I'll fix up this one.

Commit:
Gives status effects autism

Xhuis commented:
I reverted to the old system. It's wonky and _barely_ functional, but it works for the time being and doesn't cause any gameplay issues. I'll post back once I've done some more testing.

Commit:
Fixes and stuff

Xhuis commented:
* Inath-neq's Endowment has an icon now and properly renders its text.

This is merge-ready!

Iamgoofball commented:
@Xhuis you didnt fucking add a status effect removal proc

Iamgoofball commented:
or a "has status effect" proc

Xhuis commented:
Actually, I did, but they refused to work so I implemented the barebones version to work on later.
